src/audio/detect_test.c

```diff
@@ -0,0 +1,553 @@
+		return -EINVAL;
+	}
+
+	if (cd->load_memory) {
+		rfree(cd->load_memory);
+		cd->load_memory = NULL;
+	}
+
+	return 0;
+}
+
+static int alloc_mem_load(struct comp_data *cd, uint32_t size)
+{
+	if (!cd) {
+		trace_keyword_error("alloc_mem_load() error: invalid cd");
+		return -EINVAL;
+	}
+
+	free_mem_load(cd);
+
+	if (!size)
+		return 0;
+
+	cd->load_memory = rballoc(RZONE_BUFFER, SOF_MEM_CAPS_RAM, size);
+
+	if (!cd->load_memory) {
+		trace_keyword_error("alloc_mem_load() alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero(cd->load_memory, size);
+
+	return 0;
+}
+
+static void detect_test_free_parameters(struct comp_data *cd)
+{
+	rfree(cd->config);
+	cd->config = NULL;
+}
+
+static struct comp_dev *test_keyword_new(struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *keyword;
+	struct sof_ipc_comp_process *ipc_keyword =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *kd;
+
+	trace_keyword("test_keyword_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_keyword->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KEYWORD, ipc_keyword->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		      COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	keyword = (struct sof_ipc_comp_process *)&dev->comp;
+	memcpy(keyword, ipc_keyword, sizeof(struct sof_ipc_comp_process));
+
+	kd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*kd));
+
+	if (!kd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, kd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void test_keyword_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_free()");
+
+	detect_test_free_parameters(cd);
+	free_mem_load(cd);
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int test_keyword_params(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+
+	dev->params.frame_fmt = config->frame_fmt;
+
+	/* detect comp uses mono */
+	dev->params.channels = 1;
+
+	dev->frame_bytes = comp_frame_bytes(dev);
+
+	/* calculate period size based on config */
+	cd->period_bytes = dev->frames * dev->frame_bytes;
+
+	return 0;
+}
+
+/**
+ * \brief Sets keyword control command.
+ * \param[in,out] dev keyword base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int test_keyword_ctrl_set_cmd(struct comp_dev *dev,
+				     struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		trace_keyword("test_keyword_ctrl_set_cmd(), "
+			      "SOF_CTRL_CMD_SWITCH: %d",
+			      cdata->chanv[0].value);
+
+		cd->detect_enabled = cdata->chanv[0].value;
+		/* TODO: Should we reset the cd->detected flag here?
+		 * Or is it better to leave it for the comp re-start?
```

Comment:
@slawblauciak better to reset here as well as this is when the detection is enabled